File: HackerRank/Algorithm/Greedy/flowers/flowers.py
#!/usr/bin/python

import fileinput

class buddy:
	
	def __init__(self):
		self.flowers = []

	def cost(self):
		total = 0
		mult = 1
		for i in self.flowers:
			total += i*mult
			mult += 1
		return total

def main():
	num_flowers, num_friends = raw_input().split()
	num_flowers = int(num_flowers)
	num_friends = int(num_friends)
	cost = [int(o) for o in raw_input().split()]

	# Reverse sort
	cost.sort(reverse = True)
	friends = []
	for i in range(num_friends):
		friends.append(buddy())

	for c in range(len(cost)):
		friends[c%num_friends].flowers.append(cost[c])

	total_cost = 0
	for bud in friends:
		total_cost += bud.cost()

	print total_cost
	# for i in cost:
	# 	print i

if __name__ == '__main__':
	main()

File: HackerRank/Algorithm/Greedy/flowers/flowers.cc
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;

// https://www.hackerrank.com/challenges/flowers

class buddy {
	public:
		vector<long long> flowers;

	long long cost() {
		long long mult = 1;
		int total = 0;
		for(auto i: flowers) {
			total += i*mult;
			mult++;
		}

		return total;
	}

	void print() {
		cout << "\n";
		for(auto i : flowers) {
			cout << "\t" << i;
		}
		cout << "\n";
	}
};

int main() {

	long long num_flowers, num_friends;
	vector<long long> cost;
	cin >> num_flowers >> num_friends;

	for(long long i = 0; i < num_flowers; i++) {
		long long c;
		cin >> c;
		cost.push_back(c);
	}

	sort(cost.begin(), cost.end(), greater<long long>());

	vector<buddy*> friends;
	for(long long i = 0; i < num_friends; i++) {
		friends.push_back(new buddy());
	}

	long long index;
	while(cost.size() > 0) {
		long long flower_cost = cost[0];
		friends[index%num_friends]->flowers.push_back(flower_cost);
		cost.erase(cost.begin());
		index++;
	}

	long long total_cost = 0;
	for(auto i : friends) {
		total_cost += i->cost();
	}

	cout << total_cost;

	return 0;
}